moon/controller.py

```diff
@@ -0,0 +1,383 @@
+            if min_dist(robot.scan[60:210]) < 1.2 and not robot.inException:
+                raise LidarCollisionException()
+
+    # context manager functions
+    def __enter__(self):
+        self.callback = self.robot.register(self.update)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.robot.unregister(self.callback)
+
+
+class SpaceRoboticsChallenge(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        bus.register("desired_speed", "artf_xyz", "artf_cmd", "pose2d", "pose3d", "driving_recovery", "request")
+
+        self.monitors = []
+        self.last_position = None
+        self.max_speed = 1.0  # oficial max speed is 1.5m/s
+        self.max_angular_speed = math.radians(60)
+
+        self.min_safe_dist = config.get('min_safe_dist', 2.0)
+        self.dangerous_dist = config.get('dangerous_dist', 1.5)
+        self.safety_turning_coeff = config.get('safety_turning_coeff', 0.8)
+        scan_subsample = config.get('scan_subsample', 1)
+        obstacle_influence = config.get('obstacle_influence', 0.8)
+        direction_adherence = math.radians(config.get('direction_adherence', 90))
+        self.local_planner = LocalPlanner(
+                obstacle_influence=obstacle_influence,
+                direction_adherence=direction_adherence,
+                max_obstacle_distance=4.0,
+                scan_subsample=scan_subsample,
+                max_considered_obstacles=100)
+
+        self.origin = None  # unknown initial position
+        self.origin_quat = quaternion.identity()
+        self.start_pose = None
+        self.yaw_offset = None
+        self.yaw, self.pitch, self.roll = 0, 0, 0
+        self.xyz = (0, 0, 0)  # 3D position for mapping artifacts
+        self.xyz_quat = [0, 0, 0]
+        self.offset = (0, 0, 0)
+        self.use_gimbal = False # try to keep the camera on level as we go over obstacles
+
+        self.brakes_on = False
+        self.camera_change_triggered_time = None
+        self.camera_angle = 0.0
+
+        self.score = 0
+        self.current_driver = None
+        
+        self.inException = False
+
+        self.last_status_timestamp = None
+        
+        self.virtual_bumper = None
+        self.rand = Random(0)
+
+    def register(self, callback):
+        self.monitors.append(callback)
+        return callback
+
+    def unregister(self, callback):
+        assert callback in self.monitors
+        self.monitors.remove(callback)
+        
+    def send_speed_cmd(self, speed, angular_speed):
+        if self.virtual_bumper is not None:
+            self.virtual_bumper.update_desired_speed(speed, angular_speed)
+        self.bus.publish('desired_speed', [round(speed*1000), round(math.degrees(angular_speed)*100)])
+
+    def send_request(self, cmd):
+        """Send ROS Service Request form a single place"""
+        self.publish('request', cmd)
+        while True:
```

Comment:
Is there any timeout or how do we guarantee this will not stay there forever?

Also, how do we know the response that will arrive belongs to the expected request? Are we aiming to be thread-safe with this?

Reply:
hmm - what did former code in case of timeout? The code in master can parametrize the timeout in config (see #514 ) but it keeps trying for-ever. The logic can be forward here and the answer would be "None" or "timeout" instead of expected data?? Please add comment to https://github.com/robotika/osgar/pull/514/files#diff-1acffe2908dd3554c46d8ff3d6f6c86eR91

Reply:
I will add comment from e-mail "how can we recognize what answer is to my query"? This "version 0" works only for a single customer, i.e. it was his query. My goal was with minimal code change extend existing functionality with logging (so we can debug what was happening after 40 minutes simulation, or "overnight parallel simulation batch"). With @zwn we discussed that if we want to propagate `timeout` it should be something you cannot send, i.e. maybe another dedicated message for `msgpack`?? For now I would survive even with string "__timeout__" and handle it in this only calling function.